# Task
BUSCA LA COLUMNA PIB PROMEDIO POR PAÍS EN LA TABLA PIB BUENO4 Y TOMANDO EN CUENTA LA COLUMNA Country name UNE ESA COLUMNA A LA TABLA TABLAFINALCONLAQUEPUEDOHACERCOSAS TOMANDO EN CUENTA LA COLUMNA Country Name 2

Here is all the data you need:
"TABLAFINALCONLAQUEPUEDOHACERCOSAS.csv"
"PIB BUENO4.csv"

## Data loading

### Subtask:
Load the two CSV files into pandas DataFrames.


**Reasoning**:
Load the two CSV files into pandas DataFrames, handling potential encoding issues.



In [ ]:
import pandas as pd

try:
    df_final = pd.read_csv('TABLAFINALCONLAQUEPUEDOHACERCOSAS.csv', encoding='latin-1')
except UnicodeDecodeError:
    try:
        df_final = pd.read_csv('TABLAFINALCONLAQUEPUEDOHACERCOSAS.csv', encoding='utf-8')
    except Exception as e:
        print(f"Error loading TABLAFINALCONLAQUEPUEDOHACERCOSAS.csv: {e}")
        df_final = None

try:
    df_pib = pd.read_csv('PIB BUENO4.csv', encoding='latin-1')
except UnicodeDecodeError:
    try:
        df_pib = pd.read_csv('PIB BUENO4.csv', encoding='utf-8')
    except Exception as e:
        print(f"Error loading PIB BUENO4.csv: {e}")
        df_pib = None

if df_final is not None:
    display(df_final.head())
if df_pib is not None:
    display(df_pib.head())

,Country,Country Name 2,Country Code,1960 Población Total 0-14,1964 Población Total 0-14,1968 Población Total 0-14,1972 Población Total 0-14,1976 Población Total 0-14,1980 Población Total 0-14,1984 Población Total 0-14,...,1996_oro,2000_oro,2004_oro,2008_oro,2012_oro,2016_oro,2020_oro,2024_oro,total_gold_medals,promedio_de_medallas_de_oro_por_país_1960-2024
0,Aruba,Aruba,ABW,23349,23536,22817,21249,18742,17210,17017,...,0,0,0,0,0,0,0,0,0,0.0
1,Afghanistan,Afghanistan,AFG,3789907,4162207,4671648,5261617,5909060,6065314,5239339,...,0,0,0,0,0,0,0,0,0,0.0
2,Angola,Angola,AGO,2159095,2354637,2574082,2856140,3215857,3634050,4177523,...,0,0,0,0,0,0,0,0,0,0.0
3,Albania,Albania,ALB,662424,766535,854827,921354,951736,968929,1003721,...,0,0,0,0,0,0,0,0,0,0.0
4,Andorra,Andorra,AND,2570,3642,5206,6701,7761,8365,8984,...,0,0,0,0,0,0,0,0,0,0.0


,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2015,2016,2017,2018,2019,2020,2021,2022,2024,PIB PROMEDIO POR PAÍS
0,Aruba,ABW,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,27458.225330,27441.529660,28440.051960,30082.127640,31096.205070,22855.932320,27200.061080,30559.533530,33984.790620,21553.56576
1,NaN,AFE,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,186.132432,186.947182,197.408105,225.447007,209.005786,226.883067,...,1479.564123,1329.777824,1520.171298,1538.924188,1493.780445,1344.080962,1522.590088,1628.024526,1659.515290,830.9099831
2,Afganistán,AFG,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,565.569730,522.082216,525.469771,491.337221,496.602504,510.787063,356.496214,357.261153,415.707417,415.5761757
3,NaN,AFW,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,121.938353,127.452629,133.825452,139.006714,148.547736,155.563837,...,1845.767804,1616.843198,1560.162999,1703.896392,1783.654365,1664.249176,1747.840549,1777.235012,1568.724310,868.1627855
4,Angola,AGO,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,3213.902611,1807.952941,2437.259712,2538.591391,2189.855714,1449.922867,1925.874661,2929.694455,2308.159767,1730.932861


## Data exploration

### Subtask:
Explore the loaded dataframes `df_final` and `df_pib` to understand their structure and identify potential issues.


**Reasoning**:
Examine `df_final` and `df_pib` to understand their structure, identify missing values, and summarize key characteristics of relevant columns, as per the instructions.



In [ ]:
# Examine df_final
print("Shape of df_final:", df_final.shape)
display(df_final.head())
print("\nData types of df_final columns:\n", df_final.dtypes)
print("\nMissing values in df_final:\n", df_final.isnull().sum())

print("\nSummary of 'Country Name 2' column in df_final:")
print("Data type:", df_final['Country Name 2'].dtype)
print("Number of missing values:", df_final['Country Name 2'].isnull().sum())
print("Unique values:", df_final['Country Name 2'].nunique())
print("Most frequent values:\n", df_final['Country Name 2'].value_counts().head())


# Examine df_pib
print("\n\nShape of df_pib:", df_pib.shape)
display(df_pib.head())
print("\nData types of df_pib columns:\n", df_pib.dtypes)
print("\nMissing values in df_pib:\n", df_pib.isnull().sum())

print("\nSummary of 'Country name' column in df_pib:")
print("Data type:", df_pib['Country Name'].dtype)
print("Number of missing values:", df_pib['Country Name'].isnull().sum())
print("Unique values:", df_pib['Country Name'].nunique())
print("Most frequent values:\n", df_pib['Country Name'].value_counts().head())

print("\nSummary of 'PIB PROMEDIO POR PAÍS' column in df_pib:")
print("Data type:", df_pib['PIB PROMEDIO POR PAÍS'].dtype)
print("Number of missing values:", df_pib['PIB PROMEDIO POR PAÍS'].isnull().sum())
print("Descriptive statistics:\n", df_pib['PIB PROMEDIO POR PAÍS'].describe())

Shape of df_final: (207, 147)


,Country,Country Name 2,Country Code,1960 Población Total 0-14,1964 Población Total 0-14,1968 Población Total 0-14,1972 Población Total 0-14,1976 Población Total 0-14,1980 Población Total 0-14,1984 Población Total 0-14,...,1996_oro,2000_oro,2004_oro,2008_oro,2012_oro,2016_oro,2020_oro,2024_oro,total_gold_medals,promedio_de_medallas_de_oro_por_país_1960-2024
0,Aruba,Aruba,ABW,23349,23536,22817,21249,18742,17210,17017,...,0,0,0,0,0,0,0,0,0,0.0
1,Afghanistan,Afghanistan,AFG,3789907,4162207,4671648,5261617,5909060,6065314,5239339,...,0,0,0,0,0,0,0,0,0,0.0
2,Angola,Angola,AGO,2159095,2354637,2574082,2856140,3215857,3634050,4177523,...,0,0,0,0,0,0,0,0,0,0.0
3,Albania,Albania,ALB,662424,766535,854827,921354,951736,968929,1003721,...,0,0,0,0,0,0,0,0,0,0.0
4,Andorra,Andorra,AND,2570,3642,5206,6701,7761,8365,8984,...,0,0,0,0,0,0,0,0,0,0.0



Data types of df_final columns:
 Country                                            object
Country Name 2                                     object
Country Code                                       object
1960 Población Total 0-14                           int64
1964 Población Total 0-14                           int64
                                                   ...   
2016_oro                                            int64
2020_oro                                            int64
2024_oro                                            int64
total_gold_medals                                   int64
promedio_de_medallas_de_oro_por_país_1960-2024    float64
Length: 147, dtype: object

Missing values in df_final:
 Country                                           0
Country Name 2                                    0
Country Code                                      0
1960 Población Total 0-14                         0
1964 Población Total 0-14                         0
           

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2015,2016,2017,2018,2019,2020,2021,2022,2024,PIB PROMEDIO POR PAÍS
0,Aruba,ABW,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,27458.225330,27441.529660,28440.051960,30082.127640,31096.205070,22855.932320,27200.061080,30559.533530,33984.790620,21553.56576
1,NaN,AFE,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,186.132432,186.947182,197.408105,225.447007,209.005786,226.883067,...,1479.564123,1329.777824,1520.171298,1538.924188,1493.780445,1344.080962,1522.590088,1628.024526,1659.515290,830.9099831
2,Afganistán,AFG,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,565.569730,522.082216,525.469771,491.337221,496.602504,510.787063,356.496214,357.261153,415.707417,415.5761757
3,NaN,AFW,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,121.938353,127.452629,133.825452,139.006714,148.547736,155.563837,...,1845.767804,1616.843198,1560.162999,1703.896392,1783.654365,1664.249176,1747.840549,1777.235012,1568.724310,868.1627855
4,Angola,AGO,PIB per cápita (US$ a precios actuales),NY.GDP.PCAP.CD,NaN,NaN,NaN,NaN,NaN,NaN,...,3213.902611,1807.952941,2437.259712,2538.591391,2189.855714,1449.922867,1925.874661,2929.694455,2308.159767,1730.932861



Data types of df_pib columns:
 Country Name              object
Country Code              object
Indicator Name            object
Indicator Code            object
1960                     float64
                          ...   
2020                     float64
2021                     float64
2022                     float64
2024                     float64
PIB PROMEDIO POR PAÍS     object
Length: 69, dtype: object

Missing values in df_pib:
 Country Name               2
Country Code               0
Indicator Name             0
Indicator Code             0
1960                     116
                        ... 
2020                       8
2021                       8
2022                      11
2024                      23
PIB PROMEDIO POR PAÍS      0
Length: 69, dtype: int64

Summary of 'Country name' column in df_pib:
Data type: object
Number of missing values: 2
Unique values: 264
Most frequent values:
 Country Name
Zambia             2
Zimbabwe           2
Yemen, Rep. del    

## Data preparation

### Subtask:
Prepare the data for merging by renaming a column in `df_pib` and handling missing values in the join key column ("Country Name 2") of both DataFrames.


**Reasoning**:
Rename the 'Country Name' column in `df_pib` to 'Country Name 2' and handle the missing values in the 'Country Name 2' column of both dataframes. Since there are only 2 missing values in 'Country Name' column of `df_pib`, and no missing values in 'Country Name 2' column of `df_final`, dropping those two rows in `df_pib` seems a reasonable choice.



In [ ]:
# Rename the 'Country Name' column in df_pib
df_pib = df_pib.rename(columns={'Country Name': 'Country Name 2'})

# Drop rows with missing values in 'Country Name 2' column of df_pib
df_pib_cleaned = df_pib.dropna(subset=['Country Name 2'])

# Verify the changes
print("Number of missing values in 'Country Name 2' column of df_final:", df_final['Country Name 2'].isnull().sum())
print("Number of missing values in 'Country Name 2' column of df_pib:", df_pib_cleaned['Country Name 2'].isnull().sum())

Number of missing values in 'Country Name 2' column of df_final: 0
Number of missing values in 'Country Name 2' column of df_pib: 0


## Data wrangling

### Subtask:
Merge the two dataframes, `df_final` and `df_pib_cleaned`, based on the common column "Country Name 2".


**Reasoning**:
Merge the two dataframes based on the "Country Name 2" column using an inner join, then check for and remove duplicates.



In [ ]:
# Merge the two dataframes
df_merged = pd.merge(df_final, df_pib_cleaned, on='Country Name 2', how='inner')

# Check the shape of the merged dataframe
print("Shape of df_merged:", df_merged.shape)
display(df_merged.head())

# Remove duplicate rows
df_merged = df_merged.drop_duplicates(keep='first')

# Display the updated dataframe
print("\nShape of df_merged after removing duplicates:", df_merged.shape)
display(df_merged.head())

Shape of df_merged: (209, 215)


,Country,Country Name 2,Country Code_x,1960 Población Total 0-14,1964 Población Total 0-14,1968 Población Total 0-14,1972 Población Total 0-14,1976 Población Total 0-14,1980 Población Total 0-14,1984 Población Total 0-14,...,2015,2016,2017,2018,2019,2020,2021,2022,2024,PIB PROMEDIO POR PAÍS
0,Aruba,Aruba,ABW,23349,23536,22817,21249,18742,17210,17017,...,27458.225330,27441.529660,28440.051960,30082.127640,31096.205070,22855.932320,27200.061080,30559.533530,33984.790620,21553.56576
1,Angola,Angola,AGO,2159095,2354637,2574082,2856140,3215857,3634050,4177523,...,3213.902611,1807.952941,2437.259712,2538.591391,2189.855714,1449.922867,1925.874661,2929.694455,2308.159767,1730.932861
2,Albania,Albania,ALB,662424,766535,854827,921354,951736,968929,1003721,...,3952.803574,4124.055390,4531.032207,5365.489347,5460.428237,5370.778623,6413.283286,6846.426143,8575.171134,2527.403368
3,Andorra,Andorra,AND,2570,3642,5206,6701,7761,8365,8984,...,38654.934720,40129.819200,40672.994330,42819.774580,41257.804580,37361.090070,42425.699680,42414.059010,46812.448450,24569.27808
4,United Arab Emirates,Emiratos Árabes Unidos,ARE,49608,55093,62228,108874,165845,278468,394215,...,43534.997270,41326.401820,42340.591710,45689.857000,44251.454140,37173.875410,43360.021100,49899.065300,49040.694750,33121.00846



Shape of df_merged after removing duplicates: (205, 215)


,Country,Country Name 2,Country Code_x,1960 Población Total 0-14,1964 Población Total 0-14,1968 Población Total 0-14,1972 Población Total 0-14,1976 Población Total 0-14,1980 Población Total 0-14,1984 Población Total 0-14,...,2015,2016,2017,2018,2019,2020,2021,2022,2024,PIB PROMEDIO POR PAÍS
0,Aruba,Aruba,ABW,23349,23536,22817,21249,18742,17210,17017,...,27458.225330,27441.529660,28440.051960,30082.127640,31096.205070,22855.932320,27200.061080,30559.533530,33984.790620,21553.56576
1,Angola,Angola,AGO,2159095,2354637,2574082,2856140,3215857,3634050,4177523,...,3213.902611,1807.952941,2437.259712,2538.591391,2189.855714,1449.922867,1925.874661,2929.694455,2308.159767,1730.932861
2,Albania,Albania,ALB,662424,766535,854827,921354,951736,968929,1003721,...,3952.803574,4124.055390,4531.032207,5365.489347,5460.428237,5370.778623,6413.283286,6846.426143,8575.171134,2527.403368
3,Andorra,Andorra,AND,2570,3642,5206,6701,7761,8365,8984,...,38654.934720,40129.819200,40672.994330,42819.774580,41257.804580,37361.090070,42425.699680,42414.059010,46812.448450,24569.27808
4,United Arab Emirates,Emiratos Árabes Unidos,ARE,49608,55093,62228,108874,165845,278468,394215,...,43534.997270,41326.401820,42340.591710,45689.857000,44251.454140,37173.875410,43360.021100,49899.065300,49040.694750,33121.00846


## Summary:

### 1. Q&A

The task was to merge the "PIB PROMEDIO POR PAÍS" column from the "PIB BUENO4.csv" file with the "TABLAFINALCONLAQUEPUEDOHACERCOSAS.csv" file, using "Country Name" as the key.  This was accomplished by renaming the "Country Name" column in the PIB dataframe to "Country Name 2" to match the corresponding column in the other dataframe, then merging the two dataframes based on the common column.  Duplicate rows were subsequently removed.

### 2. Data Analysis Key Findings

*   The "PIB PROMEDIO POR PAÍS" column in `df_pib` initially had an incorrect data type (object) and contained "#DIV/0!" values, suggesting errors in the average GDP calculation.
*   Two rows with missing "Country Name" values were removed from `df_pib`, creating `df_pib_cleaned` with 268 rows.
*   An inner join of `df_final` (207 rows) and `df_pib_cleaned` resulted in a merged dataframe (`df_merged`) with 209 rows.
*   Four duplicate rows were removed from `df_merged`, resulting in a final dataframe with 205 rows.

### 3. Insights or Next Steps

*   Investigate the "#DIV/0!" values in the "PIB PROMEDIO POR PAÍS" column and correct the GDP calculation errors in the original "PIB BUENO4.csv" file.
*   Explore the reason for the four duplicate rows in the merged dataframe and correct the source data if necessary.  Consider using a different join type if appropriate (e.g. left join) in case you want to keep all rows from the original dataframe.


In [ ]:
df_merged.to_excel('CONLAQUEPUEDOHACERCOSAS.xlsx', index=False)